src/java.base/share/classes/java/security/PEMDecoder.java

```diff
@@ -0,0 +1,486 @@
+                        // If the KeyFactory decoded an algorithm-specific
+                        // encodings, look for the public key again.  This
+                        // happens with EC and SEC1-v2 encoding
+                        X509EncodedKeySpec spec = new X509EncodedKeySpec(
+                            p8.getPubKeyEncoded(), algo);
+                        yield new KeyPair(getKeyFactory(algo).
+                            generatePublic(spec), p8);
+                    } else {
+                        // No public key, return the private key.
+                        yield d;
+                    }
+                }
+                case Pem.ENCRYPTED_PRIVATE_KEY -> {
+                    if (password == null) {
+                        yield new EncryptedPrivateKeyInfo(decoder.decode(
+                            pem.pem()));
+                    }
+                    yield new EncryptedPrivateKeyInfo(decoder.decode(pem.pem())).
+                        getKey(password.getPassword());
+                }
+                case Pem.CERTIFICATE, Pem.X509_CERTIFICATE,
+                     Pem.X_509_CERTIFICATE -> {
+                    CertificateFactory cf = getCertFactory("X509");
+                    yield (X509Certificate) cf.generateCertificate(
+                        new ByteArrayInputStream(decoder.decode(pem.pem())));
+                }
+                case Pem.X509_CRL, Pem.CRL -> {
+                    CertificateFactory cf = getCertFactory("X509");
+                    yield (X509CRL) cf.generateCRL(
+                        new ByteArrayInputStream(decoder.decode(pem.pem())));
+                }
+                case Pem.RSA_PRIVATE_KEY -> {
+                    KeyFactory kf = getKeyFactory("RSA");
+                    yield kf.generatePrivate(
+                        RSAPrivateCrtKeyImpl.getKeySpec(decoder.decode(
+                            pem.pem())));
+                }
+                default -> pem;
+            };
+        } catch (GeneralSecurityException | IOException e) {
+            throw new IllegalArgumentException(e);
+        }
+    }
+
+    /**
+     * Decodes and returns a {@link DEREncodable} from the given {@code String}.
+     *
+     * <p> This method reads the {@code String} until PEM data is found
+     * or the end of the {@code String} is reached.  If no PEM data is found,
+     * an {@code IllegalArgumentException} is thrown.
+     *
+     * <p> This method returns a Java API cryptographic object,
+     * such as a {@code PrivateKey}, if the PEM type is supported.
+     * Any non-PEM data preceding the PEM header is ignored by the decoder.
+     * If no cryptographic object is found, a {@link PEMRecord} will be
+     * returned containing the type identifier and Base64-encoded data.
+     * Any non-PEM data preceding the PEM header will be stored in
+     * {@code leadingData}.
+     *
+     * <p>If no PEM data is found, an {@code IllegalArgumentException} is
```

Comment:
Duplicated sentence. Or do you mean to say the same exception is thrown if decoding fails?